crates/types/src/bindings.rs

```diff
@@ -1,118 +1,146 @@
-    fn from(event: &Zenith::BlockSubmitted) -> Zenith::BlockHeader {
-        Zenith::BlockHeader {
-            rollupChainId: event.rollupChainId,
-            sequence: event.sequence,
-            confirmBy: event.confirmBy,
-            gasLimit: event.gasLimit,
-            rewardAddress: event.rewardAddress,
-            blockDataHash: event.blockDataHash,
-        }
+// returns a BlockHeader from a BlockSubmitted event with the given host block number
+#[allow(dead_code)]
+pub(crate) const fn header_from_block_submitted(
+    event: &Zenith::BlockSubmitted,
+    host_block_number: U256,
+) -> Zenith::BlockHeader {
+    Zenith::BlockHeader {
+        rollupChainId: event.rollupChainId,
+        hostBlockNumber: host_block_number,
+        gasLimit: event.gasLimit,
+        rewardAddress: event.rewardAddress,
+        blockDataHash: event.blockDataHash,
     }
 }
 
-impl Zenith::ZenithEvents {
-    /// Get the chain ID of the event (discarding high bytes), returns `None`
-    /// if the event has no associated chain id.
-    pub const fn rollup_chain_id(&self) -> Option<u64> {
+// Passage types
+impl Copy for Passage::EnterConfigured {}
+impl Copy for Passage::Withdrawal {}
+impl Copy for Passage::OnlyTokenAdmin {}
+impl Copy for Passage::Enter {}
+impl Copy for Passage::EnterToken {}
+
+impl Clone for Passage::PassageEvents {
+    fn clone(&self) -> Self {
         match self {
-            Zenith::ZenithEvents::BlockSubmitted(inner) => Some(inner.rollup_chain_id()),
-            Zenith::ZenithEvents::Enter(inner) => Some(inner.rollup_chain_id()),
-            Zenith::ZenithEvents::Transact(inner) => Some(inner.rollup_chain_id()),
-            _ => None,
+            Self::Enter(inner) => Self::Enter(*inner),
+            Self::Withdrawal(inner) => Self::Withdrawal(*inner),
+            Self::EnterConfigured(inner) => Self::EnterConfigured(*inner),
+            Self::EnterToken(inner) => Self::EnterToken(*inner),
+            Self::Transact(inner) => Self::Transact(inner.clone()),
         }
     }
 }
 
-impl Zenith::BlockSubmitted {
+impl Passage::Enter {
     /// Get the chain ID of the event (discarding high bytes), returns `None`
     /// if the event has no associated chain id.
     pub const fn rollup_chain_id(&self) -> u64 {
         self.rollupChainId.as_limbs()[0]
     }
 }
 
-impl Zenith::Enter {
-    /// Get the chain ID of the event (discarding high bytes), returns `None`
-    /// if the event has no associated chain id.
+impl Passage::PassageErrors {}
+impl Passage::OnlyTokenAdmin {}
+impl Passage::EnterToken {}
+impl Passage::EnterConfigured {}
+impl Passage::Enter {}
+impl Passage::DisallowedEnter {}
+
+impl Passage::Transact {
     pub const fn rollup_chain_id(&self) -> u64 {
         self.rollupChainId.as_limbs()[0]
     }
+    pub const fn sender(&self) -> Address {
+        self.sender
+    }
+    pub const fn to(&self) -> Address {
+        self.to
+    }
+    pub fn data(&self) -> Bytes {
+        self.data.clone()
```

Comment:
instead of cloning, we can return a borrow here and allow the caller to decide whether to clone or not 

```
-> &Bytes { &self.data }
```